### PR TITLE
feat(node): upstream rollup node p2p

### PR DESCRIPTION
**Pull request by Pangssu**
# Description

Upstream to Optimism's commit af736c12e36e93fb30efdccd8d663f445b0bc7a8 because Rollup Node P2P's peer discovery is not working, and changed Rollup Node CLI package to v2

### CODE REVIEW

**Review by kangsorang — APPROVED** [review]
That's a lot of code, but it's localized to a specific module, so it seems okay.

File: FORK.md [hunk 1]
```diff
@@ -5,3 +5,7 @@
 - tag: `v1.0.3`
 - differences
   - [Differences from Optimism Bedrock](./specs/differences-from-optimism-bedrock.md)
+
+### Rollup Node P2P
+
+- https://github.com/ethereum-optimism/optimism/commit/af736c12e36e93fb30efdccd8d663f445b0bc7a8
```
Comment (seolaoh) [inline FORK.md:11]:
```suggestion
- <https://github.com/ethereum-optimism/optimism/commit/af736c12e36e93fb30efdccd8d663f445b0bc7a8>
```

File: components/node/metrics/metrics.go [hunk 2]
```diff
@@ -10,17 +10,19 @@ import (
...
 	pb "github.com/libp2p/go-libp2p-pubsub/pb"
 	libp2pmetrics "github.com/libp2p/go-libp2p/core/metrics"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/collectors"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
 
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rpc"
+
```
Comment (seolaoh) [inline components/node/metrics/metrics.go:22]: Please run `goimports-reviser -rm-unused ./...`

File: specs/rollup-node-p2p.md [hunk 3]
```diff
@@ -1,7 +1,7 @@
 # Rollup-node P2P interface
 
 The [rollup node](./rollup-node.md) has an optional peer-to-peer (P2P) network service to improve the latency between
-the view of proposers and the rest of the network by bypassing the L1 in the happy case,
+the view of sequencers and the rest of the network by bypassing the L1 in the happy case,
```
Comment (seolaoh) [inline specs/rollup-node-p2p.md:4]: Please change `sequencer` to `proposer` in this file.
```suggestion
the view of proposers and the rest of the network by bypassing the L1 in the happy case,
```

File: specs/rollup-node-p2p.md [hunk 4]
```diff
@@ -81,38 +81,36 @@ Common representations of network identity:
...
 - chain ID (`unsigned varint`)
 - fork ID (`unsigned varint`)
 
-Note that DiscV5 is a shared [DHT (Distributed Hash Table)][dht]: the L1 consensus and execution nodes,
+Note that DiscV5 is a shared DHT (Distributed Hash Table): the L1 consensus and execution nodes,
 as well as testnet nodes, and even external IOT nodes, all communicate records in this large common DHT.
 This makes it more difficult to censor the discovery of node records.
 
-The discovery process in Kroma is a pipeline of node records:
+The discovery process in Optimism is a pipeline of node records:
```
Comment (seolaoh) [inline specs/rollup-node-p2p.md:100]:
```suggestion
The discovery process in Kroma is a pipeline of node records:
```

File: specs/rollup-node-p2p.md [hunk 5]
```diff
@@ -81,38 +81,36 @@ Common representations of network identity:
 
 #### Structure
 
-The [Ethereum Node Record (ENR)][enr] for a Kroma rollup node must contain the following values, identified by unique
-keys:
+The Ethereum Node Record (ENR) for an Optimism rollup node must contain the following values, identified by unique keys:
```
Comment (seolaoh) [inline specs/rollup-node-p2p.md:84]:
```suggestion
The Ethereum Node Record (ENR) for an Kroma rollup node must contain the following values, identified by unique keys:
```

File: specs/rollup-node-p2p.md [hunk 6]
```diff
@@ -311,7 +309,7 @@ TODO: GossipSub per-topic scoring to fine-tune incentives for ideal propagation
 
 ## Req-Resp
 
-The kroma-node implements a similar request-response encoding for its sync protocols as the L1 ethereum Beacon-Chain.
+The op-node implements a similar request-response encoding for its sync protocols as the L1 ethereum Beacon-Chain.
```
Comment (seolaoh) [inline specs/rollup-node-p2p.md:312]: Please change `op-node` to `kroma-node` in this file.
```suggestion
The kroma-node implements a similar request-response encoding for its sync protocols as the L1 ethereum Beacon-Chain.
```

File: utils/service/util.go [hunk 7]
```diff
@@ -3,16 +3,78 @@ package service
...
 }
 
+// PrefixEnvVarV2 adds a prefix to the environment variable,
+// and returns the env-var wrapped in a slice for usage with urfave CLI v2.
+func PrefixEnvVarV2(prefix, suffix string) []string {
+	return []string{prefix + "_" + suffix}
+}
+
+// ValidateEnvVars logs all env vars that are found where the env var is
+// prefixed with the supplied prefix (like OP_BATCHER) but there is no
```
Comment (seolaoh) [inline utils/service/util.go:30]:
```suggestion
// prefixed with the supplied prefix (like KROMA_BATCHER) but there is no
```

File: utils/service/util_test.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+	"github.com/stretchr/testify/require"
+	"github.com/urfave/cli/v2"
+)
+
+func TestCLIFlagsToEnvVars(t *testing.T) {
+	flags := []cli.Flag{
+		&cli.StringFlag{
+			Name:    "test",
+			EnvVars: []string{"OP_NODE_TEST_VAR"},
```
Comment (seolaoh) [inline utils/service/util_test.go:14]: Please change `OP_` to `KROMA_` in this file.
```suggestion
			EnvVars: []string{"KROMA_NODE_TEST_VAR"},
```

File: e2e/system_test.go [hunk 9]
```diff
@@ -719,6 +721,24 @@ func TestSystemMockP2P(t *testing.T) {
...
 	require.NoError(t, err)
+
+	syncerPeerID := sys.RollupNodes["syncer"].P2P().Host().ID()
+	check := func() bool {
+		proposerBlocksTopicPeers := sys.RollupNodes["proposer"].P2P().GossipOut().BlocksTopicPeers()
+		return slices.Contains[peer.ID](proposerBlocksTopicPeers, syncerPeerID)
+	}
+
+	// poll to see if the verifier node is connected & meshed on gossip.
+	// Without this verifier, we shouldn't start sending blocks around, or we'll miss them and fail the test.
```
Comment (seolaoh) [inline e2e/system_test.go:732]:
```suggestion
	// poll to see if the validator node is connected & meshed on gossip.
	// Without this validator, we shouldn't start sending blocks around, or we'll miss them and fail the test.
```

File: e2e/system_test.go [hunk 10]
```diff
@@ -719,6 +721,24 @@ func TestSystemMockP2P(t *testing.T) {
...
+	backOffStrategy := backoff.Exponential()
+	for i := 0; i < 10; i++ {
+		if check() {
+			break
+		}
+		time.Sleep(backOffStrategy.Duration(i))
+	}
+	require.True(t, check(), "verifier must be meshed with proposer for gossip test to proceed")
```
Comment (seolaoh) [inline e2e/system_test.go:740]:
```suggestion
	require.True(t, check(), "validator must be meshed with proposer for gossip test to proceed")
```

**Review by seolaoh — APPROVED** [review]
LGTM

**Review by 0xHansLee — APPROVED** [review]

**Pull request merged**